File: app/schemas/backtest.py
"""Backtest Pydantic schemas."""

from datetime import date
from typing import Any

from pydantic import BaseModel, Field


class BacktestCreate(BaseModel):
    """Create backtest request."""

    strategy_id: int
    etf_code: str
    start_date: date
    end_date: date
    initial_capital: float = 100000.0
    commission_rate: float = 0.001
    slippage_rate: float = 0.001
    position_size: float = 1.0
    risk_free_rate: float = 0.02


class BacktestMetrics(BaseModel):
    """Backtest performance metrics."""

    initial_capital: float
    final_nav: float
    total_return: float
    annualized_return: float
    max_drawdown: float
    sharpe_ratio: float
    win_rate: float
    trade_count: int
    avg_win: float
    avg_loss: float
    trading_days: int
    commission_rate: float = 0.001
    slippage_rate: float = 0.001
    position_size: float = 1.0
    risk_free_rate: float = 0.02
    # New (quant P0-9) — None means "not computable" rather than 0
    sortino_ratio: float | None = None
    calmar_ratio: float | None = None
    var_95: float | None = None
    cvar_95: float | None = None
    max_drawdown_duration: int | None = None
    annualization_factor: int = 252
    # Execution-config echoes
    execution_price_model: str = "open"
    market: str = "cn_a"
    apply_friction: bool = True


class BacktestTrade(BaseModel):
    """Backtest trade record."""

    entry_date: str
    exit_date: str | None = None
    entry_price: float
    exit_price: float
    side: str
    pnl: float
    pnl_pct: float


class BacktestResponse(BaseModel):
    """Backtest response."""

    id: int
    strategy_id: int
    start_date: str
    end_date: str
    metrics: dict[str, Any]
    trades: list[dict[str, Any]]
    daily_nav: list[dict[str, Any]]
    signals: list[dict[str, Any]]
    config_snapshot: dict[str, Any] | None = None
    created_at: str | None = None


class BacktestListItem(BaseModel):
    """Backtest list item."""

    id: int
    strategy_id: int
    etf_code: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    metrics: dict[str, Any]
    trade_count: int
    created_at: str | None = None


class BacktestListResponse(BaseModel):
    """Backtest list response."""

    items: list[BacktestListItem]


# ---------------------------------------------------------------------------
# Cross-sectional backtest schema (quant P1)
# ---------------------------------------------------------------------------


class CrossSectionalBacktestCreate(BacktestCreate):
    """Create cross-sectional (multi-symbol) backtest request.

    The legacy single-code path is preserved via the parent class.
    Provide ``etf_codes`` to switch into cross-sectional mode.
    """

    etf_codes: list[str] | None = None
    # Allocation per symbol is fixed at ``1 / len(etf_codes)`` (equal
    # weights, daily rebalance to the long side of signals). No
    # configurable field for that today.


# ---------------------------------------------------------------------------
# Parameter optimization schemas (quant P1)
# ---------------------------------------------------------------------------


class OptimizeRequest(BaseModel):
    """Request body for ``POST /backtests/optimize``.

    Accepts the same inputs as ``BacktestCreate`` plus a ``grid`` map
    specifying the value lists to sweep over. ``etf_code`` and
    ``etf_codes`` are both honoured — when ``etf_codes`` is provided
    the optimizer runs a cross-sectional sweep.
    """

    strategy_type: str = Field(..., description="Registered strategy identifier")
    etf_code: str | None = None
    etf_codes: list[str] | None = None
    base_params: dict[str, Any] = Field(default_factory=dict)
    grid: dict[str, list[Any]] = Field(
        default_factory=dict,
        description="Mapping of param_name -> [v1, v2, ...]",
    )
    start_date: date
    end_date: date
    initial_capital: float = 100000.0
    commission_rate: float = 0.001
    slippage_rate: float = 0.001
    position_size: float = 1.0
    risk_free_rate: float = 0.02
    execution_price_model: str = "open"
    market: str = "cn_a"
    apply_friction: bool = True
    top_n: int = Field(10, ge=1, le=100, description="Number of Pareto-optimal rows to return")


class OptimizeResponse(BaseModel):
    """Response body for ``POST /backtests/optimize``."""

    top_n: list[dict[str, Any]] = Field(
        default_factory=list,
        description="Up to ``top_n`` Pareto-optimal candidates by Sharpe.",
    )
    full_sweep_size: int = 0
    strategy_type: str
    grid_keys: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Composite strategy schema (quant P1)
# ---------------------------------------------------------------------------


class CompositeComponent(BaseModel):
    """One component inside a composite strategy."""

    type: str = Field(..., description="Component strategy_type")
    params: dict[str, Any] = Field(default_factory=dict)
    weight: float = 1.0


class CompositeStrategyConfig(BaseModel):
    """Composite strategy configuration.

    Used as the body of ``POST /backtests/composite``. The endpoint
    routes the request through the generic ``CompositeStrategy`` in
    ``app.strategies.base`` — register your components by their
    ``strategy_type`` strings.
    """

    components: list[CompositeComponent] = Field(..., min_length=1)
    aggregation: str = Field(
        "weighted",
        description="weighted | vote | unanimous",
        pattern="^(weighted|vote|unanimous)$",
    )
    holding_period: int = 20

    def as_params(self) -> dict[str, Any]:
        """Render into the dict accepted by ``run_backtest``/``Strategy``."""
        return {
            "components": [
                {"type": c.type, "params": c.params, "weight": c.weight}
                for c in self.components
            ],
            "aggregation": self.aggregation,
            "holding_period": self.holding_period,
        }


class CompositeBacktestRequest(CompositeStrategyConfig):
    """Composite backtest request: composite config + backtest envelope."""

    etf_code: str = Field(..., description="Single-instrument target")
    start_date: date
    end_date: date
    initial_capital: float = 100000.0
    commission_rate: float = 0.001
    slippage_rate: float = 0.001
    position_size: float = 1.0
    risk_free_rate: float = 0.02
    execution_price_model: str = "open"
    market: str = "cn_a"
    apply_friction: bool = True

File: web/src/types/backtest.ts
export interface BacktestMetrics {
  initial_capital: number;
  final_nav: number;
  total_return: number;
  annualized_return: number;
  max_drawdown: number;
  sharpe_ratio: number;
  win_rate: number;
  trade_count: number;
  avg_win: number;
  avg_loss: number;
  trading_days: number;
}

export interface BacktestTrade {
  entry_date: string;
  exit_date?: string;
  entry_price: number;
  exit_price: number;
  side: string;
  pnl: number;
  pnl_pct: number;
}

export interface BacktestNAV {
  date: string;
  nav: number;
}

export interface BacktestSignal {
  date: string;
  signal_type: string;
  price: number;
}

export interface Backtest {
  id: number;
  strategy_id: number;
  start_date: string;
  end_date: string;
  metrics: Record<string, any>;
  trades: BacktestTrade[];
  daily_nav: BacktestNAV[];
  signals: BacktestSignal[];
  config_snapshot?: Record<string, any>;
  created_at?: string;
}

export interface BacktestListItem {
  id: number;
  strategy_id: number;
  start_date?: string;
  end_date?: string;
  metrics: Record<string, any>;
  trade_count: number;
  created_at?: string;
}

export interface BacktestListResponse {
  items: BacktestListItem[];
}

export interface BacktestCreate {
  strategy_id: number;
  etf_code: string;
  start_date: string;
  end_date: string;
  initial_capital?: number;
  commission_rate?: number;
  slippage_rate?: number;
  position_size?: number;
}

export interface AttributionEffect {
  sector?: string;
  allocation?: number;
  selection?: number;
  interaction?: number;
  total?: number;
}

export interface AttributionSummary {
  allocation_pct?: number;
  selection_pct?: number;
  interaction_pct?: number;
}

export interface AttributionTradeStats {
  total_trades?: number;
  winning_trades?: number;
  losing_trades?: number;
  win_rate?: number;
  avg_return?: number;
}

export interface AttributionResponse {
  backtest_id?: number;
  total_return: number;
  benchmark_return: number;
  excess_return: number;
  effects?: AttributionEffect[];
  summary?: AttributionSummary;
  trade_stats?: AttributionTradeStats;
}
